org.jacoco.core/src/org/jacoco/core/internal/Java9Support.java

```diff
@@ -0,0 +1,98 @@
+/*******************************************************************************
+ * Copyright (c) 2009, 2016 Mountainminds GmbH & Co. KG and Contributors
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Evgeny Mandrikov - initial API and implementation
+ *
+ *******************************************************************************/
+package org.jacoco.core.internal;
+
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.Opcodes;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * Patching for Java 9 classes, so that ASM can read them.
+ */
+public final class Java9Support {
+
+	public static final int V1_9 = Opcodes.V1_8 + 1;
+
+	private Java9Support() {
+	}
+
+	/**
+	 * Copy of {@link ClassReader#readClass(InputStream, boolean)}.
+	 */
+	public static byte[] readClass(final InputStream is, boolean close)
+			throws IOException {
+		if (is == null) {
+			throw new IOException("Class not found");
+		}
+		try {
+			byte[] b = new byte[is.available()];
```

Comment:
This code has quite some complexity. We should either add tests or use more high level APIs. E.g. the input can be easily copied into a ByteArrayOutputStream which then most of the copy magic for us (needs a little bit more memory allocations though).